File: src/main/java/BinarySearch.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class BinarySearch {

    public static int findSubArray(int[] array, int sum) {

        boolean found = false;
        int subArraySum = 0;
        int start = 0;
        int end = array.length -1;
        for (int i = 0; i < array.length; i++) {
            subArraySum += array[i];
            if(subArraySum < sum){
                continue;
            }else if(subArraySum > sum){
                subArraySum = 0;
                start = i+1;
            }else{
                end  = i;
                found = true;
                break;
            }
        }
        if(found)  return end -start + 1 ;
        else return 0;
    }

    public static int findRadius(int[] houses, int[] heaters) {

        Arrays.sort(houses);
        Arrays.sort(heaters);
        int[] housesNearest = new int[houses.length];
        for (int i = 0; i < houses.length; i++) {
            housesNearest[i] = searchNearest(heaters , houses[i]);
        }
        Arrays.sort(housesNearest);
        return housesNearest[housesNearest.length -1 ];
    }

    public static int searchNearest(int[] heaters, int house) {

        int nearestDistance = Math.abs(heaters[0] - house);

        for (int i = 0; i < heaters.length; i++) {
            int distance = Math.abs(heaters[i] - house);
            if (distance <= nearestDistance) {
                nearestDistance = distance;
            }
        }
        return nearestDistance;
    }

    public static int getMinimumUniqueSum(List<Integer> arr) {

        Set<Integer> myset = new HashSet<>();
        for(int i  : arr){
            myset.add(i);
        }

        int sum = 0;
        for (Integer num : myset) {
            sum+= num;
        }
        return sum;

    }

    public static void main(String[] args) {

        int array []= new int[]{1,1,1,1,1,1,1,1} ;
       // System.out.println("findSubArray "+ findSubArray(array,11));
        //int houses []= new int[]{1,5} ; int heaters[] = new int[]{2};
        //System.out.println("findRadius : "+ findRadius(houses,heaters));

        //int matrtix[][] = new int[][]{{1 ,2} , {3,4}}; //, target = 13;
        //System.out.println("searchMatrix = " + searchMatrix(matrtix ,4));
        //System.out.println("guess Num : " + guessNumber(2126753390));

        /*int[] houses = new int[]{1,2,3};
        int[] heaters = new int[]{2};
        System.out.println("find radius " + findRadius(houses,heaters));*/

       /* int[] arr = new int[]{ -2,0,10,-19,4,6,-8} ;
        System.out.println("checkIfExist :" + checkIfExist(arr)) ;
*/
        //System.out.println("firstBadVersion = " + firstBadVersion(1000));

        /*int[] f = new int []{ 0,0,0,1,1,1,1 ,1,1};
        System.out.println("first index = " + BSfindFirst(f , 0 , f.length , 0));
        System.out.println("last index = " + BSfindLast(f , 0 , f.length , 0));

        int[] f1 = new int []{1,1,1,1,0,0,0,0,0,0};
        System.out.println("last index  desc= " + BSfindLastDecreasing(f1 , 0 , f1.length , 1));*/
    }

    public static boolean searchMatrix(int[][] matrix, int target) {

        int start = 0 ; int end = matrix.length -1 ; int mid = 0;

        //Single element matrix
        if(matrix.length ==1)
            return Arrays.binarySearch(matrix[0],0 , matrix[0].length , target) >=0 ;

        while(start <= end ){

            mid = (start + end) / 2;

            if(Arrays.binarySearch(matrix[mid],0 , matrix[mid].length  , target) >=0 ){
                return true;
            }else {
                // target should be in this row ?
                if(target > matrix[mid][0] && target < matrix[mid][matrix[mid].length -1 ]) return false;

                //adjust start and end
                if(target > matrix[mid][0]) start = mid +1  ;

                else if (target < matrix[mid][matrix[mid].length -1 ]) end = mid -1 ;
            }
        }
        return false;

    }



    public static int guessNumber(int n) {
        long start = 1 ; long end = n ;
        while (start <= end){
            long mid = (start + end) / 2; //
            int midInt = new Integer(mid + "");
            //1 : Your guess is lower than the number I picked
            if (guess(midInt) == 1)
                start = mid + 1 ;

            //-1 : Your guess is higher than the number I picked
            else if (guess(midInt) == -1)
                end = mid-1 ;

            else
                return midInt;
        }
        return -1;
    }

    private static int guess(long guess) {
        if(guess < 1702766719) return 1;
        if(guess > 1702766719) return -1;
        return 0;
    }

    public  static boolean checkIfExist(int[] arr) {
        //sort array arr = [1,7,11,12 ,14]
        Arrays.sort(arr);
        for (int i = 0; i < arr.length; i++) {
            if(findDoubleElementInArray(arr,arr[i]*2 ,i )) {
                return true;
            }
        }
        return false;
    }
    private static boolean findDoubleElementInArray(int[] arr, int i , int index) {
        int foundIndex  = Arrays.binarySearch(arr, i)  ;
        return  foundIndex >=0 && foundIndex != index;
    }

    public static int firstBadVersion(int n) {
        int start = 1 ; int end = n ; int badVersion = 0;
        while(start <= end) {
            int mid = start + (end-start) / 2;
            if(isBadVersion(mid)) {
                System.out.println("bad version = " + mid);
                badVersion = mid;
                end = mid -1;
            }
            else {
                start = mid+1;
            }
            System.out.println(mid + " " + start + " " + end );
        }
        return badVersion;
    }

    private static boolean isBadVersion(int version) {
        return version < 1000;
    }

    public static int BSfindLast(int[] f ,int start, int end, int val) {
        while(start < end) {
            int mid = start + (end-start+1) / 2;
            if(f[mid] < val)		start= mid + 1;
            else if(f[mid] > val)	end  = mid - 1;
            else start = mid;
        }
        return start;
    }

    public static int BSfindLastDecreasing(int[] f ,int start, int end, int val) {
        int index = 0 ;
        while(start < end && (end - start > 2)) {
            int mid = start + (end-start) / 2;
            if(f[mid] < val)		end= mid -1;
            else if(f[mid] > val)	start  = mid+1 ;
            else start = mid;
        }

        for(int i=start ; i< end ; i++){
            if(f[i] == val) index = i;
        }
        return  index;
    }

    public  static int BSfindFirst(int[] f,int start, int end, int val) {

        while(start < end) {
            int mid = start + (end-start) / 2;
            int item = f[mid];
            if(item < val)		start= mid + 1;
            else if(item > val)	end  = mid - 1;
            else end  = mid;
        }

        return start;
    }

    public  static  int binary_search(int A[], int key, int imin, int imax){

        // continue searching while [imin,imax] is not empty
        while (imin < imax){

            // calculate the midpoint for roughly equal partition
            int imid = (imin + imax) / 2; //(imin + imax) / 2

            // determine which subarray to search
            if (A[imid] < key)

                // change min index to search upper subarray
                imin = imid + 1;

            else if (A[imid] > key)
                // change max index to search lower subarray
                imax = imid - 1;

            else
                // key found at index imid
                return imid;
        }

        // key not found
        return -1;
    }


    public  static  int binary_search_lower_or_higher(int A[], int key, int imin, int imax){

        // continue searching while [imin,imax] is not empty
        while (imin < imax){
            // calculate the midpoint for roughly equal partition
            int imid = (imin + imax) / 2; //(imin + imax) / 2
            // determine which subarray to search
            if (A[imid] < key)
                // change min index to search upper subarray
                imin = imid + 1;

            else if (A[imid] > key)
                // change max index to search lower subarray
                imax = imid - 1;

            else
                // key found at index imid
                return 0;
        }

        // key not found
        return -1;
    }
}

File: src/main/java/sessions/BinarySearch.cpp
		// [Binary Search]

		/*

			+ How do you find a page number in a book?

					1-5000, 3000

				- First part, last part, or the target



			+ We have a search space that can guide us if we will continue 
				in the right position or in the left position or we will not continue searching



			+ The classical exapmle of having a sorted array 
				and we search for a value in it we want to get the index of it

				L = 1 3 4 6 8 9 11     X = 11

				   Compare X to 6. It's smaller. Repeat with L = 1 3 4.
				   Compare X to 3. It's bigger. Repeat with L = 4.
				   Compare X to 4. It's equal. We're done, we found X.

				- The idea of binary search is that we choose the middle of the search space
					each time, as the middle is the optimum choise we can do each time, from here the word binary came.

				- Other examples are Number guessing game, and Word list.

		*/



		// [Recursive]
		// A straightforward implementation of binary search is recursive. 
		// The initial call uses the indices of the entire array to be searched. 
		// The procedure then calculates an index midway between the two indices, 
		// 	determines which of the two subarrays to search, and then does a recursive call to search that subarray. 
		// Each of the calls is tail recursive, so a compiler need not make a new stack frame for each call. 
		// The variables imin and imax are the lowest and highest inclusive indices that are searched.

		int binary_search(int A[], int key, int imin, int imax){

		  // test if array is empty
		  if (imax < imin)

		    // set is empty, so return value showing not found
		    return KEY_NOT_FOUND;

		  else{

		      // calculate midpoint to cut set in half
		      int imid = midpoint(imin, imax); //(imin + imax) / 2
		 
		      // three-way comparison

		      if (A[imid] > key)
		        // key is in lower subset
		        return binary_search(A, key, imin, imid-1);

		      else if (A[imid] < key)
		        // key is in upper subset
		        return binary_search(A, key, imid+1, imax);

		      else
		        // key has been found
		        return imid;
		    }
		}



		//  [Iterative]
		//	The binary search algorithm can also be expressed iteratively with two index limits that progressively narrow the search range

			int binary_search(int A[], int key, int imin, int imax){

			  // continue searching while [imin,imax] is not empty
			  while (imin < imax){

			      // calculate the midpoint for roughly equal partition
			      int imid = midpoint(imin, imax); //(imin + imax) / 2
			 
			      // determine which subarray to search
			      if (A[imid] < key)

			        // change min index to search upper subarray
			        imin = imid + 1;

			      else if (A[imid] > key)
			        // change max index to search lower subarray
			        imax = imid - 1;

			      else
			        // key found at index imid
			        return imid;
			    }

			  // key not found
			  return KEY_NOT_FOUND;
			}





		// The order of binary search is O(log N) - Because the search space is decreasing by a division factor







		// Java implements binary search by


		// + Using Arrays.binarySearch()
			
			int index = Arrays.binarySearch(sortedArray, value);

		// A sortedArray and an int value, which is to be searched in the array of integers, are passed as arguments to the binarySearch method of the Java Arrays class.


		// + Using Collections.binarySearch()

			int index = Collections.binarySearch(sortedList, value);

		// A sortedList & an Integer value, which is to be searched in the list of Integer objects, are passed as arguments to the binarySearch method of the Java Collections class.





		/*
		 * Binary Search works only if the search space can guide you. That is from any position you know to which direction to go?
		 * 	In more theorical words: binary search can be used if and only if for all x in S, p(x) implies p(y) for all y > x
		 *
		 * Typically works in increasing or decreasing sequence. Some times works in bitonic sequence (increase then decrease or reverse) (1 7 9 15 6 4 2)
		 */




		/*
		 * 	1- Instead of an index in array, we could have function F(int idx) that calculate some value.
		 *
		 * 		E.g. we have an increasing function F (e.g. Fibonacci 1 1 2 3 5 8 13 ...) what is index of 21?
		 *
		 *
		 * 	2- In many problems, function F is boolean function could(int val).
		 *
		 * 		E.g. If we are given # of sandwiches to eat during trip. Using them you could reach your destination or die from starving as food ended.
		 * 			What is MIN # of sandwiches to finish trip and still alive?
		 *
		 * 		in this case space is like  0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1
		 *
		 * 			That is e.g. given up to 6 sandwiches u die, but more than that you could pass.
		 *
		 * 		Sometimes space is reverse 1 1 1 1 1 0 0 0 0 0 0
		 *
		 * 		Sometimes you are asked for first 1 (or) zero and some times you are asked for the last.
		 *
		 * 		For more understanding for first and last: Imagine array: 1 3 4 5 9 9 9 12 15 ...
		 * 			Find 1st idx of 9? Find last idx of 9?
		 * 			So it is not just about finding, but which position? any? first? last?
		 *
		 * 		Final note: remember (e.g. in 0 0 0 0 0 0 1 1 1 1 1) the last zero followed by the first one. So you could be asked to last, and solve it usign first.
		 */




		// start + (end - start) / 2 = (start + end) / 2

		int BSfindFirst(int start, int end, int val) {

			while(start < end) {

				int mid = start + (end-start) / 2;

				if(f(mid) < val)		start= mid + 1;
				else if(f(mid) > val)	end  = mid - 1;
				else end  = mid;
			}

			return start;
		}

		int BSfindLast(int start, int end, int val) {

			while(start < end) {

				int mid = start + (end-start) / 2;

				if(f(mid) < val)		start= mid + 1;
				else if(f(mid) > val)	end  = mid - 1;
				else start = mid;
			}

			return start;
		}



		// A typical scenario is infinite loop when a start or end = mid. To fix that
		// check to use: mid = start + (end-start)/2;	OR		 mid = start + (end-start+1)/2;
		// check any 2 consecutive numbers: E.g. 1, 2


		// If you forgot this trick, use Halawany technique xD, just work till range be > small number (e.g. 5) and then move the remain steps sequentially.









		// [Binary Search in the double domain] //


		// One of great WAs sources is value of EPS. Make sure you achieve required precision

		double BS_double(double start, double end, double val) {

			while(fabs(end - start) > EPS) {	// iterate 200-600 iteration

				double mid = (start + end) / 2;

				if (F(mid) < val)	start = mid;
				else end  = mid;
			}

			return start;
		}


		// Halawany technique

		double binary_search_double(double start, double end, double target){

		    for (int i = 0 ; i < 600; i ++){ // iterate 200-600 iteration

		        double mid = (start + end) / 2;

		        if (F(mid) < target) start = mid;
		        else end = mid;
		    } 

		    return start;
		}


		// [Problems]

		// https://leetcode.com/problems/search-insert-position/
		// https://leetcode.com/problems/search-a-2d-matrix/
		// https://leetcode.com/problems/guess-number-higher-or-lower/
		// https://leetcode.com/problems/first-bad-version/
		// https://leetcode.com/problems/check-if-n-and-its-double-exist/
		// https://leetcode.com/problems/heaters/
		https://leetcode.com/problems/minimum-number-of-days-to-make-m-bouquets/